includes/class-wc-checkout.php

```diff
@@ -686,27 +686,32 @@ public function get_posted_data() {
 			}
 
 			foreach ( $fieldset as $key => $field ) {
+				if ( isset( $_POST[ $key ] ) ) { // phpcs:disable WordPress.Security.NonceVerification.Missing
+					$value = wp_unslash( $_POST[ $key ] ); // phpcs:ignore WordPress.Security.ValidatedSanitizedInput.InputNotSanitized
+				} elseif ( isset( $field['default'] ) ) {
+					$value = $field['default'];
+				} else {
+					$value = '';
+				}
+
 				$type = sanitize_title( isset( $field['type'] ) ? $field['type'] : 'text' );
 
-				// phpcs:disable WordPress.Security.NonceVerification.Missing
 				switch ( $type ) {
```

Comment:
There's a problem here: this `switch` statement is executed always, but it should execute only when there's no value set. Imagine for example a checkbox that defaults to 0 and is left unchecked, with the current code it will be stored as a 1.

To really mimic the previous logic the `switch` should be executed only when there's no value set:

```php
if('' !== $value) {
  switch(...) {...}
}
```

Reply:
Goot catch. done!  42bc356

Reply:
```php
$value = 0 === $value || "0" === $value || false === $value ? '' : 1;
```
Adding that line in that place will cause if we add a non-required input with `'foo'` as default value it will saved as `"1"` even if the input was left unchecked since we are not checking for `"foo"` or any other other values.

[Looking](https://github.com/woocommerce/woocommerce/blob/f8340f80cbd9365ba5f4bb69a3fe5680e7a767b3/includes/wc-template-functions.php#L2807) [at](https://github.com/WordPress/WordPress/blob/2e762721b1e0cb634b85b315f58c82bf8f42e1b5/wp-includes/general-template.php#L4753-L4755) [the code](https://github.com/WordPress/WordPress/blob/2e762721b1e0cb634b85b315f58c82bf8f42e1b5/wp-includes/general-template.php#L4819-L4820) it seems the only way to mark a checkbox in the UI is to use a value that translates to `"1"` after string conversion. So I think it's better if we use the same logic here.

```php
$value = (string) $value === "1" ? 1 : '';
```

There was a small issue I've found though. When I added an optional checkbox with `"1"` as default value. It was saved even though the user unchecked the input. It made sense because the script didn't know if the user unchecked the input or it was never not presented to the user taking its default value instead. That's why I moved the line to the top and also added a conditional to know if the user was presented with the checkout form. 

I've tested my changes with many cases and can confirm that:

#### When UI is presented to the user (checkout page).
- [x] Checked inputs are saved as "1" no matter its default value or if they are required or optional.
- [x] Optional inputs are not saved if not checked on submit
- [x] Optional inputs are saved as "1" if marked on submit

#### When UI is not presented to the user (product page)
  - [x] Required inputs are saved as "1" if it has a default value that translates to "1"
  - [x] Optional inputs are saved as "1" if it has a default value that translates to "1" 
  - [x] Error message is displayed for required inputs with a default value that does not translate to "1"
  - [x] Optional inputs are not saved regardless if they have a default value or not.

<details>

```php
$fields['billing']['billing_checkbox11'] = array(
  'type' => 'checkbox',
  'required' => true,
  'label' => 'billing_checkbox11',
  'default' => '1'
);
$fields['billing']['billing_checkbox12'] = array(
  'type' => 'checkbox',
  'required' => true,
  'label' => 'billing_checkbox12',
  'default' => true
);
$fields['billing']['billing_checkbox13'] = array(
  'type' => 'checkbox',
  'required' => true,
  'label' => 'billing_checkbox13',
  'default' => 1
);
$fields['billing']['billing_checkbox14'] = array(
  'type' => 'checkbox',
  'required' => true,
  'label' => 'billing_checkbox14',
  'default' => '0'
);
$fields['billing']['billing_checkbox15'] = array(
  'type' => 'checkbox',
  'required' => true,
  'label' => 'billing_checkbox15',
  'default' => 0
);
$fields['billing']['billing_checkbox16'] = array(
  'type' => 'checkbox',
  'required' => true,
  'label' => 'billing_checkbox16',
  'default' => ''
);
$fields['billing']['billing_checkbox17'] = array(
  'type' => 'checkbox',
  'required' => true,
  'label' => 'billing_checkbox17',
  'default' => ' '
);
$fields['billing']['billing_checkbox18'] = array(
  'type' => 'checkbox',
  'required' => true,
  'label' => 'billing_checkbox18',
  'default' => 'foo'
);
$fields['billing']['billing_checkbox19'] = array(
  'type' => 'checkbox',
  'label' => 'billing_checkbox19',
  'default' => 'foo'
);
$fields['billing']['billing_checkbox20'] = array(
  'type' => 'checkbox',
  'label' => 'billing_checkbox20',
  'default' => '1'
);
```

</details>

---

```php
$is_checkout_page = isset( $_REQUEST['woocommerce-process-checkout-nonce'] );
```

This is probably not the best way to identify if the request came from the checkout page and/or the user was presented with the checkout form. Please let me know if is there a better way to know that.

Reply:
I think that what we need is simply to exclude checkboxes from the default mechanism:

```php
if ( isset( $_POST[ $key ] ) && '' !== $_POST[ $key ] ) {
	$value = wp_unslash($_POST[$key]);
} elseif ( isset( $field['default'] ) && 'checkbox' !== $type ) {
	$value = $field['default'];
} else {
	$value = '';
}
```

This way it will work whether the default is checked or unchecked and whether the user changes the value or not.

"But what about mandatory checkboxes that are unchecked by default?", you may ask. We'll, consider what's the purpose of such a checkbox: it's usually to have the user agree on some terms and conditions. It doesn't make sense to add such a confirmation field in scenarios where the user doesn't have an opportunity to explicitly make a choice.

Additionally, you'll see the first line of my suggestion is `if ( isset( $_POST[ $key ] ) && '' !== $_POST[ $key ] )`. I have added the comparison against an empty string because otherwise, an empty text field will cause `isset( $_POST[ $key ] )` to be true (the key is set and has a value of `''`), but semantically that's an unset field and thus the default should be used.

Does that make sense?

Reply:
@Konamiman Yeah, I think it makes sense. I've tested it and it's working as expected

Reply:
Also do I need to add an entry to the `changelog.txt`? Where do I put it?

Reply:
No, you should just place an appropriate one-line summary under "Changelog entry" in the pull request description, we handle `changelog.txt` entries ourselves at the time of preparing a new release.